daemon/api_system_secureboot.go

```diff
@@ -0,0 +1,159 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2024 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package daemon
+
+import (
+	"encoding/base64"
+	"encoding/json"
+	"net/http"
+
+	"github.com/snapcore/snapd/overlord/auth"
+	"github.com/snapcore/snapd/overlord/fdestate"
+)
+
+var systemSecurebootCmd = &Command{
+	// TODO GET returning whether secure boot is relevant for the system?
+
+	Path: "/v2/system/secureboot",
+	POST: postSystemSecurebootAction,
+	WriteAccess: interfaceRootAccess{
+		// TODO should this be only allowed for services with fwupd on slot
+		// side?
+		Interfaces: []string{"fwupd"},
+	},
+}
+
+func postSystemSecurebootAction(c *Command, r *http.Request, user *auth.UserState) Response {
+	contentType := r.Header.Get("Content-Type")
+
+	switch contentType {
+	case "application/json":
+		return postSystemSecurebootActionJSON(c, r)
+	default:
+		return BadRequest("unexpected content type: %q", contentType)
+	}
+}
+
+func postSystemSecurebootActionJSON(c *Command, r *http.Request) Response {
+
+	ucred, err := ucrednetGet(r.RemoteAddr)
+	if err != nil {
+		return Forbidden("cannot obtain credentials of the caller: %v", err)
+	}
+
+	var req struct {
+		Action string           `json:"action,omitempty"`
+		Data   *json.RawMessage `json:"data"`
+	}
+
+	decoder := json.NewDecoder(r.Body)
+
+	if err := decoder.Decode(&req); err != nil {
+		return BadRequest("cannot decode request body: %v", err)
+	}
+	if decoder.More() {
+		return BadRequest("extra content found in request body")
+	}
+
+	switch req.Action {
+	case "efi-secureboot-db-update":
```

Comment:
having action and subaction is a bit unusual, also at least the startup action might be relevant to more types of fw updates than db-updates?  can't we fold things into actions:

efi-secureboot-update-startup
efi-secureboot-update-db-prepare
efi-secureboot-update-db-cleanup?

is clean perhaps also relevant to other kind of updates?